src/coinsviewcacheasync.h

```diff
@@ -0,0 +1,227 @@
+ *
+ * The cache spawns a fixed set of worker threads that fetch Coins for each input in a block.
+ * When FetchCoin() is called, the main thread waits for the corresponding coin to be fetched and returns it.
+ * While waiting, the main thread will also fetch coins to maximize parallelism.
+ *
+ * Worker threads are synchronized with the main thread using a barrier, which is used at the beginning of fetching to
+ * start the workers and at the end to ensure all workers have finished before the next block is started.
+ */
+class CoinsViewCacheAsync : public CCoinsViewCache
+{
+private:
+    //! The latest input not yet being fetched. Workers atomically increment this when fetching.
+    mutable std::atomic_uint32_t m_input_head{0};
+    //! The latest input not yet accessed by a consumer. Only the main thread increments this.
+    mutable uint32_t m_input_tail{0};
+
+    //! The inputs of the block which is being fetched.
+    struct InputToFetch {
+        //! Workers set this after setting the coin. The main thread tests this before reading the coin.
+        std::atomic_flag ready{};
+        //! The outpoint of the input to fetch;
+        const COutPoint& outpoint;
+        //! The coin that workers will fetch and main thread will insert into cache.
+        std::optional<Coin> coin{std::nullopt};
+
+        /**
+         * We only move when m_inputs reallocates during setup.
+         * We never move after work begins, so we don't have to copy other members.
+         */
+        InputToFetch(InputToFetch&& other) noexcept : outpoint{other.outpoint} {}
+        explicit InputToFetch(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
+    };
+    mutable std::vector<InputToFetch> m_inputs{};
+
+    /**
+     * The first 8 bytes of txids of all txs in the block being fetched. This is used to filter out inputs that
+     * are created and spent in the same block, since they will not be in the db or the cache.
+     * Using only the first 8 bytes is a performance improvement, versus storing the entire 32 bytes. In case of a
+     * collision of an input being spent having the same first 8 bytes as a txid of a tx elsewhere in the block,
+     * the input will not be fetched in the background. The input will still be fetched later on the main thread.
+     */
+    std::vector<uint64_t> m_txids{};
+
+    //! DB coins view to fetch from.
+    const CCoinsView& m_db;
+
+    /**
+     * Similar to CCoinsViewCache::GetCoin, but it does not mutate internally.
+     * Therefore safe to call from any thread once inside the barrier.
+     */
+    std::optional<Coin> GetCoinWithoutMutating(const COutPoint& outpoint) const
+    {
+        auto coin{static_cast<CCoinsViewCache*>(base)->GetPossiblySpentCoinFromCache(outpoint)};
+        if (!coin) coin = m_db.GetCoin(outpoint);
+        if (coin && !coin->IsSpent()) [[likely]] return coin;
```

Comment:
I don't mind the `[[likely]]` attributes (it helps with review as well), but I know others have strong feeling against using them.

Reply:
The first link shows a benefit to using them, and that is now in our codebase.
The second and third links reference [this blog post](https://blog.aaronballman.com/2020/08/dont-use-the-likely-or-unlikely-attributes/) which has crazy convoluted usages. I don't find that blog post convincing. The usages here are straightforward and in very hot paths. They are only used in paths where we know a valid block will always or never go into (aside from BIP30).